### PR TITLE
TYP: Typing changes for ExtensionArray.astype

### PR DESCRIPTION
**Pull request by Dr-Irv**
Minimal (hopefully) typing changes for `ExtensionArray.astype` .

### CODE REVIEW

File: pandas/core/arrays/base.py
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:517]: dt64/td64 generally go to DTA/TDA

Comment (Dr-Irv) [reply pandas/core/arrays/base.py:517]: Don't think so.  When you pass a numpy dtype, you get a `np.ndarray`:
```
>>> a=pd.array([np.datetime64("2021-03-15"), np.datetime64("2021-04-05")])
>>> a
<DatetimeArray>
['2021-03-15 00:00:00', '2021-04-05 00:00:00']
Length: 2, dtype: datetime64[ns]
>>> b=a.astype(np.datetime64)
>>> b
array(['2021-03-15T00:00:00.000000000', '2021-04-05T00:00:00.000000000'],
      dtype='datetime64[ns]')
>>> type(b)
<class 'numpy.ndarray'>
```

File: pandas/core/arrays/base.py
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:524]: the docstring and the type signature are not consistent for return type

Comment (Dr-Irv) [reply pandas/core/arrays/base.py:524]: fixed in next commit

File: pandas/core/arrays/base.py
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:558]: OK, done in next commit

File: pandas/core/arrays/base.py
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:938]: OK, done in next commit.

**Comment by Dr-Irv** [comment]
@simonjayhawkins can you review, please?

**Comment by simonjayhawkins** [comment]
This is now failing mypy with master merged. can you merge master and fixup.

File: pandas/core/groupby/groupby.py
Comment (simonjayhawkins) [inline pandas/core/groupby/groupby.py:1500]: this returns vals.view(np.int8). so np.ndarray is correct?

File: pandas/core/internals/blocks.py
Comment (simonjayhawkins) [inline pandas/core/internals/blocks.py:1354]: this is just a allow-redefinition error.

rather than ignoring, do

```
values: ArrayLike = self.values
```

on L1350

File: pandas/core/arrays/base.py
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:522]: The defaults are not needed in the overloads

```suggestion
    def astype(self, dtype: NpDtype, copy: bool = ...) -> np.ndarray:
```

File: pandas/core/arrays/base.py
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:529]: need to change the one-liner in the docstring to be consistent

File: pandas/_testing/asserters.py
Comment (simonjayhawkins) [inline pandas/_testing/asserters.py:405]: does adding an overload for `Type[str | float | int | complex | bool |object]` fix these false positives?

Comment (simonjayhawkins) [reply pandas/_testing/asserters.py:405]: or maybe move those into NpDtype from Dtype in pandas._typing.

File: pandas/core/arrays/base.py
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:526]: I think we could add an overload for ExtensionDtype and this overload could be str for the union return type if type_t[Union[str, float, int, complex, bool, object] is moved as suggested? 

the accepted strings for numpy is already known? so we should eventually be able to remove the union return type.

File: pandas/core/arrays/base.py
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:953]: see other comments

File: pandas/core/groupby/groupby.py
Comment (Dr-Irv) [reply pandas/core/groupby/groupby.py:1500]: Yes.  Changed it back.

File: pandas/core/internals/blocks.py
Comment (Dr-Irv) [reply pandas/core/internals/blocks.py:1354]: done in next commit

File: pandas/core/arrays/base.py
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:522]: done in next commit

File: pandas/_testing/asserters.py
Comment (Dr-Irv) [reply pandas/_testing/asserters.py:405]: I changed `NpDtype`, which forced a bunch of changes elsewhere, mostly to get rid of `ignore`

File: pandas/core/arrays/base.py
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:953]: fixed in next commit

**Comment by Dr-Irv** [comment]
@simonjayhawkins  pushed changes a week ago, so looking forward to a new review.

File: pandas/core/arrays/base.py
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:543]: in docstrings i usually say "np.ndarray or ExtensionArray" instead of "ArrayLike".  do we have a policy on that?

File: pandas/core/arrays/base.py
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:544]: is StringDtype a typo?

Comment (jbrockmendel) [reply pandas/core/arrays/base.py:544]: oh i see, this is bc of the special-casing.  should be fixed by #41652

File: pandas/core/construction.py
Comment (jbrockmendel) [inline pandas/core/construction.py:591]: better to just pass np.dtype("object")

File: pandas/core/dtypes/dtypes.py
Comment (jbrockmendel) [inline pandas/core/dtypes/dtypes.py:1293]: why this change?  what prevents me from passing `int`?

File: pandas/core/internals/blocks.py
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1351]: values is specifically ExtensionArray here (hopefully mypy can figure that out on its own)

File: pandas/core/arrays/base.py
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:543]: Made change in next commit.

File: pandas/core/arrays/base.py
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:544]: clarified in next commit by saying "if dtype is StringDtype"

File: pandas/core/internals/blocks.py
Comment (Dr-Irv) [reply pandas/core/internals/blocks.py:1351]: Yes, but the issue is that a few lines later, we say `values = values.astype(object)`, and `values` is an `np.ndarray` in this case, so by declaring it as `ArrayLike`, we cover that case.

File: pandas/core/construction.py
Comment (Dr-Irv) [reply pandas/core/construction.py:591]: fixed in next commit

File: pandas/core/dtypes/dtypes.py
Comment (Dr-Irv) [reply pandas/core/dtypes/dtypes.py:1293]: Fails mypy for some reason.  Might have to do with the typing of how the function `np.dtype` is typed by numpy

**Comment by Dr-Irv** [comment]
@jbrockmendel @simonjayhawkins got marked stale due to lack of review/approval....

File: pandas/core/dtypes/dtypes.py
Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:1293]: `PandasDtype(int)` returns PandasDtype('int64')

pls let the question-asker do the "mark as resolved"

File: pandas/core/arrays/categorical.py
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py:486]: i think with dt64/td64 could return DTA/TDA?

Comment (Dr-Irv) [reply pandas/core/arrays/categorical.py:486]: Don't see that:
```python
>>> s = pd.Series(pd.to_datetime(["2021-03-15 12:05", "2021-04-05 5:10"]), dtype
="category")
>>> s
0   2021-03-15 12:05:00
1   2021-04-05 05:10:00
dtype: category
Categories (2, datetime64[ns]): [2021-03-15 12:05:00, 2021-04-05 05:10:00]
>>> s.dtype.categories.dtype
dtype('<M8[ns]')
>>> type(s.astype(s.dtype.categories.dtype).values)
<class 'numpy.ndarray'>
```

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
This branch is 260 commits behind master and if merged would cause a CI failure

```
pandas/core/arrays/period.py:346: error: unused "type: ignore" comment
```

File: pandas/_testing/asserters.py
Comment (simonjayhawkins) [inline pandas/_testing/asserters.py:403]: why is this changed?

File: pandas/_typing.py
Comment (simonjayhawkins) [inline pandas/_typing.py:128]: does this PR need to go in before #41203?

File: pandas/core/arrays/boolean.py
Comment (simonjayhawkins) [inline pandas/core/arrays/boolean.py:400]:
```suggestion
    def astype(self, dtype: NpDtype, copy: bool = ...) -> np.ndarray:
```

and elsewhere

File: pandas/core/arrays/base.py
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:521]: could this use npt.DtypeLike?

File: pandas/core/arrays/base.py
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:525]: if we pass a EA dtype object, we get a union return type?

File: pandas/core/dtypes/dtypes.py
Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py:1293]: does npt.DTypeLike work?

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @Dr-Irv 

is it worth posting the results of a test file with reveal_types for this?

File: pandas/_typing.py
Comment (Dr-Irv) [reply pandas/_typing.py:128]: No, they should be independent.  Either can go first, and once one goes in, I can do the mypy check on the other.

File: pandas/_testing/asserters.py
Comment (Dr-Irv) [reply pandas/_testing/asserters.py:403]: Because when you do `astype(int)`, you 

> is it worth posting the results of a test file with reveal_types for this?

Can you point me to an example?

**Comment by Dr-Irv** [comment]
> Thanks @Dr-Irv
> 
> is it worth posting the results of a test file with reveal_types for this?

Can you point me to an example?

File: pandas/_testing/asserters.py
Comment (Dr-Irv) [reply pandas/_testing/asserters.py:403]: Must have had to change it in a previous commit with an earlier version of numpy.  Reverted in next commit.

File: pandas/core/arrays/boolean.py
Comment (Dr-Irv) [reply pandas/core/arrays/boolean.py:400]: fixed in next commit

File: pandas/core/arrays/base.py
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:521]: > could this use npt.DtypeLike?

No.  The following code is fine with mypy
```
import numpy.typing as npt

class Foo:
    pass

def myfun(x: npt.DTypeLike):
    pass

myfun(type(Foo))
myfun("abcd")
myfun(str)
myfun(object)
myfun("str")
```
That's because `npt.DTypeLike` accepts any type or any str.  We want to separate out the `ExtensionDtype` that are passed as strings.

File: pandas/core/arrays/base.py
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:525]: Good point.  Fixed in next commit

File: pandas/core/arrays/base.py
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:521]: Never mind.  Was able to get this to work by making `astype()` accept `npt.DTypeLike`.  Created a new type `AstypeArg` to mean `ExtensionDtype | npt.DTypeLike` so that I don't have to change lots of other parts where `NpDtype` is used.

**Comment by Dr-Irv** [comment]
@simonjayhawkins ping...

**Comment by Dr-Irv** [comment]
pinging @simonjayhawkins

**Comment by jreback** [comment]
@Dr-Irv thanks for the patience. can you rebase

@simonjayhawkins if you can have a look

**Comment by Dr-Irv** [comment]
weekly ping @simonjayhawkins

**Comment by Dr-Irv** [comment]
another ping @simonjayhawkins

**Comment by Dr-Irv** [comment]
biweekly ping @simonjayhawkins

**Comment by jreback** [comment]
@Dr-Irv ok if you can merge master and make sure still passing (i think it was last) time, will merge these.

**Comment by Dr-Irv** [comment]
@jreback all green except for a Windows timeout on one of the tests

**Comment by jreback** [comment]
can you rebase again just to be sure (as merged the other)

cc @twoertwein

File: pandas/_typing.py
Comment (twoertwein) [reply pandas/_typing.py:130]: Would it make sense to use `NpDtype = npt.DTypeLike` to be consistent between numpy and pandas?

File: pandas/core/arrays/integer.py
Comment (twoertwein) [inline pandas/core/arrays/integer.py:341]: I'm not sure whether the following is possible:

```py
T = TypeVar("T", npt.DTypeLike)
def astype(self, dtype: T, copy: bool = ...) -> npt.NDArray[T]
```

Might be interesting to compare that to numpy's ndarray.astype signature.

File: pandas/_typing.py
Comment (Dr-Irv) [reply pandas/_typing.py:130]: > Would it make sense to use `NpDtype = npt.DTypeLike` to be consistent between numpy and pandas?

That's a big change, as it affects a lot more than just the `astype()` signature.  The types are different:
```python
>>> import numpy.typing as npt
>>> from pandas._typing import NpDtype
>>> NpDtype
typing.Union[str, numpy.dtype, typing.Type[typing.Union[str, float, int, complex, bool, object]]]
>>> npt.DTypeLike
typing.Union[numpy.dtype, NoneType, type, numpy.typing._dtype_like._SupportsDType[numpy.dtype], str, typing.Tuple[typing.Any, int], typing.Tuple[typing.Any, typing.Union[typing.SupportsIndex, typing.Sequence[typing.SupportsIndex]]], typing.List[typing.Any], numpy.typing._dtype_like._DTypeDict, typing.Tuple[typing.Any, typing.Any]]
```

File: pandas/core/arrays/integer.py
Comment (Dr-Irv) [reply pandas/core/arrays/integer.py:341]: > I'm not sure whether the following is possible:

Doesn't work:
```python
>>> import numpy.typing as npt
>>> from typing import TypeVar
>>> T = TypeVar("T", npt.DTypeLike)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Anaconda3\envs\pandas-dev-types\lib\typing.py", line 596, in __init__
    raise TypeError("A single constraint is not allowed")
TypeError: A single constraint is not allowed
```

Comment (twoertwein) [reply pandas/core/arrays/integer.py:341]: That's how numpy does it. It is a  bit more complex:
https://github.com/numpy/numpy/blob/1fe57b2427e20f466c254ed6a98a014d9a9e574e/numpy/__init__.pyi#L1920

Not sure whether we want to mirror that - would need to keep in sync with their private type declarations.

Comment (Dr-Irv) [reply pandas/core/arrays/integer.py:341]: > Not sure whether we want to mirror that - would need to keep in sync with their private type declarations.

And the other issue is that they define some of these private types inside of the `.pyi` files so you can't import the private types.  I think this is too much of a moving target for us to get working now until numpy's typing efforts settle down.

I also believe that numpy is defining some types for typing in Cython code, which makes things even more complex to get working.

For now, I think we should remain incremental on this and accept what I have proposed and then we can investigate how we can better type the interface between pandas and numpy.

**Comment by Dr-Irv** [comment]
@jreback all green after merge with the delete/searchsorted changes

**Comment by jreback** [comment]
thanks @Dr-Irv

**Pull request merged**
